design/open/repository-operations-error-handling.md

```diff
@@ -27,73 +27,95 @@ The proposed solution introduces 2 new elements to solve the above, in terms of
-This can be checked every time the `Repository` is accessed (trivial) and by a designated cleanup procedure (part of the above mentioned GC, or a designated scan at startup).
+A `Repository` entity with `initial` status is either being created (which will soon change the status to `active`), or already failed to create without changing status to `failed` (due to, for example, a crashed process). We will decide a `Repository` is failed if it is in `initial` state, and a sufficient time of 2 minutes, has passed from its creation attempt. In that case, the `Repository` state will be changed to `failed`, making it a candidate for deletion.
+The status set attempt will occur as part of any access (`GetRepository`) that finds the repository with a status `initial` and a creation time earlier then 2 minutes ago. Note that an attempt to create a repository, that finds a repository with the same name and `status` initial, may move it to status `failed`
+This status change will also be performed by a future designated cleanup procedure (part of the above mentioned Cleaner, or a designated scan at startup).
  
 ##  Flows
-### CreateRepository
-* Create a `Repository` entity with its given `RepositoryID`, randomly generated `UniqueID` and `State` = `initial`, under the common `graveler` partition
-  * If failed return the error (`graveler.ErrNotUnique` will be returned in case a `Repository` entity with the same `RepositoryID` exists) - entity was not created - no harm done.
-* Generate the partition key for the new `Repository`, from its `RepositoryID` and `unique_identifier`
-* Create the initial `Commit` with a `CommitID` (same as it is done today) and store it with the partition key generated above
-  * If failed - try to set the `Repository` entry state to `failed` and return the error from the KV layer
-  * If cannot set the `Repository` state to `failed` it is still consistent, as it is still in `initial` state
-* Create the default `Branch` set to the initial `CommitID` and store it with the partition key generated above
-  * If failed - try to set the `Repository` entry state to `failed` and return the error from the KV layer
-  * If cannot set the `Repository` state to `failed` it is still consistent, as it is still in `initial` state
-* Set the `Repository` state to `active`
-  * If failed - try to set the `Repository` entry state to `failed` and return the error from the KV layer
-  * If cannot set the `Repository` state to `failed` it is still consistent, as it is still in `initial` state
-* Success
+### GetRepository - **Step 1**
```

Comment:
In both step1 and step2, I think we should be proactive on failure, and try to clean up after ourselves

Reply:
Proactive cleaning (or cleaning at all) is not a requirement. This is refelcted in Step 1 where I propose (as derived from the conversation thread with Oz) the minimal work to get correctness in place.
Regarding proactive cleaning in case of failure - it (a) makes the failures longer - instead of quick error response to the user, (b) has no real benefit as you can still fail during the cleaning itself, which leads to (c) additional error handling code to handle a failure to cleanup after failure. Having all cleanup tasks in one place (i.e. designated Cleaner) will simplify the code as we will have one place to handle errors during cleanup

Reply:
We are talking specifically about create repository flow, so point (a) is irrelevant as we are just talking about deleting the initial commit.
point (b) it can, but we can still try
point (c) it can be a best effort and we don't have to handle errors at this point
The biggest problem I have with not trying to clean up is the fact that in step(1) if we fail on create branch, we will have a dangling commit that no cleanup process will be able to find

Reply:
Your point about the dangling commit (or a dangling branch too, as a failure not neccessarily means it was not created) is right, but the assumption is a repository is not an entity that is created too often, and it will fail even less frequently.